### PR TITLE
LogicMonitor Pack

### PR DESCRIPTION
**Pull request by lm-ydubler**
Hi all,
Thanks for running this incubator and I'm glad to have helped @cognifloyd to get it back online.

This is the LogicMonitor (https://www.logicmonitor.com) Pack for StackStorm and we're hoping to graduate it into the StackStorm Exchange.

That said, the owned of this pack is the "LogicMonitor" account found at this url: https://github.com/logicmonitor . Therefore, I am asking that when this pack graduates into the Exchange (for legal purposes):

1. My personal account that created this PR ("lm-ydubler") loses all privileges and connection to the LogicMonitor pack, and,

2. The LogicMonitor Github Account named "logicmonitor" or "LogicMonitor" found [here](https://github.com/logicmonitor) becomes the official owner and administrator of the pack and gains all privileges to the pack and it's repository so that it can create pull requests to update the pack, remove the pack, etc. Obviously, this would be within the boundaries and context of however you guys manage packs in the Exchange.

Please keep me in the loop and let me know if I have to do anything in this process.
Thanks and best and Merry Christmas!
Yuri

### CODE REVIEW

**Review by armab — COMMENTED** [review]
Thanks for the PR!

As noted by @nzlosh in the previous comment: https://github.com/StackStorm-Exchange/exchange-incubator/pull/169#issuecomment-954124493


* Please include a licence file. Apache 2.0 licenced code will be accepted for StackStorm packs.
* Include a changelog
* Run black over Python code with a line length of 100 characters.
* Include README.md with the description about how to use the pack and its configuration
* Include a pack icon

You can find examples of these files in other StackStorm-Exchange packs:
* https://github.com/stackStorm-exchange/stackstorm-github

Speaking of Maintenance, - we can add as a maintainer with the write/review permissions anyone (up to a group of many people) from your org willing to maintain this pack. You could review the incoming Pull Requests, approve/reject them and do the releases. We, however, can't add someone as github admin to the pack to avoid leaking secrets, sudden content override, force-pushes, deletion, or renaming, for example.

Hope that helps. We're looking forward to the new LogicMonitor pack :+1:

**Comment by lm-ydubler** [comment]
@armab

- Please include a licence file. Apache 2.0 licenced code will be accepted for StackStorm packs.
- Include a changelog
- Run black over Python code with a line length of 100 characters.
- Include README.md with the description about how to use the pack and its configuration
- Include a pack icon

Done, done, done, done, and done.

Regarding pack maintainers, please add the following people as maintainers:

- LogicMonitor , https://github.com/logicmonitor
- Yuri Dubler, https://github.com/lm-ydubler , yuri.dubler@logicmonitor.com
- Skyla Urman, skyla.urman@logicmonitor.com
- Adam Johnson, adam.johnson@logicmonitor.com
- Carlos Alvarenga, carlos.alvarenga@logicmonitor.com
- Nicholas Johnson, nicholas.johnson@logicmonitor.com

Thank you kindly and please touch back as required.

**Comment by cognifloyd** [comment]
Please merge in master. We just merged fixes for CI so it will give you real feedback.

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (cognifloyd) [inline logicmonitor/sensors/logicmonitor_sensor.py:39]: So, LogicMonitor sends the api key in the POST body. I'm guessing the LM server-side doesn't have a way to add either headers or a query param with the api key. Is that right? is that the primary motivation for creating a new webhook receiving sensor?

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (cognifloyd) [inline logicmonitor/sensors/logicmonitor_sensor.py:45]: I suggest simplifying your condition. `is <bool>` is rather unusual in python.
```suggestion
            if self._auth_enabled:
```

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (cognifloyd) [inline logicmonitor/sensors/logicmonitor_sensor.py:71]: This block of code is essentially the same in both blocks. What about:

```suggestion
                # SUCCESS : Authentication succeeded
                self._log.info(C.LOG_AUTH_SUCCEEDED)
                ret = C.RES_SUCCESS_AUTH_ENABLED
```

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (cognifloyd) [inline logicmonitor/sensors/logicmonitor_sensor.py:88]:
```suggestion
            else:

                # Auth is disabled
                self._log.info(C.LOG_AUTH_DISABLED)

                # SUCCESS - Authentication disabled.
                ret = C.RES_SUCCESS_AUTH_DISABLED

            # Finally, Inject trigger/payload & return response
            del data[
                C.API_KEY_KEY
            ]  # Remove API Key from payload before it is injected with the trigger
            self._sensor_service.dispatch(C.ALERT_TRIGGER, data)
            self._log.info(f'{C.LOG_TRIGGER_DISPATCHED}"{data}"')
            return ret
```

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (cognifloyd) [inline logicmonitor/sensors/logicmonitor_sensor.py:41]: This is a very broad except block which is generally a bad idea. I think you only care about KeyError here, right?
```suggestion
            except KeyError:
```

**Review by cognifloyd — COMMENTED** [review]
I wonder why you implemented the sensor as a separate flask server instead of reusing StackStorm's builtin webhook functionality. I've left some comments to validate what I suspect is the reasoning.

Plus a few python code suggestions in the sensor.

**Comment by cognifloyd** [comment]
OK. CI is now giving real feedback. Everything needs to pass before we can merge this.

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (lm-ydubler) [reply logicmonitor/sensors/logicmonitor_sensor.py:39]: Our first implementation was to use the builtin webhook. I don't recall what caused us to use a Flask server, but it was likely for a good reason.

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (lm-ydubler) [reply logicmonitor/sensors/logicmonitor_sensor.py:45]: This was a change requested during a code review on our side, so, will leave it as is.

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (lm-ydubler) [reply logicmonitor/sensors/logicmonitor_sensor.py:71]: There is redundancy at the cost of simplicity (removing the if statements required to modify the final message). And, for simplicity on our end, will keep the code unchanged. Changing the code would require retesting everything, more work and PRs. Will just keep it how it is. Do appreciate you taking a look though.

File: logicmonitor/sensors/logicmonitor_sensor.py
Comment (lm-ydubler) [reply logicmonitor/sensors/logicmonitor_sensor.py:41]: This is a good point but don't want to push untested changes even if they are 99.99% likely to succeed. Right now, just trying to put the pack into Exchange.

**Comment by lm-ydubler** [comment]
@cognifloyd Thank you for the feedback but will not be making changes to the pack at this time due to formal overhead required with every change. Currently just trying to get it into the Exchange. Do you need me to do anything (not code related) to get the pack into the Exchange?

You made the comment above "Please merge in master. We just merged fixes for CI so it will give you real feedback." Do I have to act on this?

**Comment by cognifloyd** [comment]
> You made the comment above "Please merge in master. We just merged fixes for CI so it will give you real feedback." Do I have to act on this?

I merged in master, so this is done.

>Thank you for the feedback but will not be making changes to the pack at this time due to formal overhead required with every change. Currently just trying to get it into the Exchange. Do you need me to do anything (not code related) to get the pack into the Exchange?

The CI tests are failing, so you need to fix the code. We won't add a pack to the exchange where CI tests are failing.

**Comment by cognifloyd** [comment]
Again, I'm sorry the CI has been broken for so long. Had it been working you would have had this feedback as soon as you created the initial PR. Thank you for helping us get it up and running again! I hope you don't have to jump through too much "formal overhead" to resolve the linting issues.

File: logicmonitor/sensors/sensor_constants.py
Comment (cognifloyd) [inline logicmonitor/sensors/sensor_constants.py:47]: You can do something like this:
```suggestion
RES_SUCCESS_AUTH_DISABLED = (
    "SUCCESS! Trigger/payload injected into ST2. "
    "AUTH IS DISABLED! Enable it in your config file!"
)
```

**Comment by lm-ydubler** [comment]
It is what it is, glad you're operational again.

Some of the code changes you requested above would require formal testing and PR because they would have changed alot. Some other changes are minor changes that can probably squeak by without testing. Either way, I will be testing the pack after making the code changes required to get it into the Exchange and I will have to make a PR in LM to update our version of the pack.

Right now, updating code to get it into the Exchange...

**Comment by lm-ydubler** [comment]
I am rushing to get the pack into the exchange but that is not really wise behavior because haste makes waste. My weekend starts tomorrow so I will be back on Monday to make sure all the changes we make to the Pack are tested and go through formal channels before it is uploaded to the Exchange. I will redo some changes such as using multiline strings (as you pointed out) in sensor_constants.py.

Do the scripts automatically add the pack to the Exchange or do I have to confirm it?
Thanks again

**Comment by cognifloyd** [comment]
I marked the PR as a draft. We won't add the pack until you click "ready for review", and a TSC member approves it (prob me).
And, no, creating the pack repo is not automated yet. I'm working on that, but I don't expect to use the automation to add this pack.

PS all the linters are green :) So, let me know when you're ready for me to take another look.

**Comment by armab** [comment]
@lm-ydubler I've created a new LogicMonitor Github group and invited you there so you could be a maintainer for the future pack.
However, for the other members, I'll need their existing github usernames, rather than emails.
Also, https://github.com/logicmonitor is an organization and not a user.

**Comment by lm-ydubler** [comment]
@cognifloyd @armab
I'll be testing the pack today one last time to make sure all the changes are functional before pushing the "Ready for review" button. I'll alert you when testing has completed. Thanks again!

**Comment by lm-ydubler** [comment]
@cognifloyd @armab The pack has been tested successfully and can be merged into the Exchange.

If we can add maintainers to the pack after it gets pushed to the Exchange, you can list me as the only maintainer of the pack for the time being with the intention of adding others as time progresses.

Thanks so much, please keep me updated as required.

**Review by cognifloyd — APPROVED** [review]

**Comment by cognifloyd** [comment]
https://github.com/StackStorm-Exchange/stackstorm-logicmonitor
https://exchange.stackstorm.org/#logicmonitor

This has been published to the exchange. :tada:

**Comment by armab** [comment]
Thanks all!

We announced it in StackStorm commication channels for better outreach.

* https://twitter.com/StackStorm/status/1481034961331363843
* https://www.linkedin.com/feed/update/urn:li:activity:6886800524469952512

@lm-ydubler If @LogicMonitor could re-share it with their community, - that would be great.
If you guys want to announce the StackStorm integration somewhere in your engineering blog (https://www.logicmonitor.com/blog), - that would be even better!

**Pull request closed without merging**
